README.md

# husk discord Bot -  V2.0

__**Updates**__:
1. Change in the maincommand `error_handler` (More efficient and it looks better)
2. Few commands are removed including: `MoveRage` - `Logger`..
3. The old helpcommand has been replaced by a  `dynamic Menu based Help command` to automate command adding.
4. The bot has become more object oriented (if we ignore d.py), you can see all new objects in `sdk/husk_sdk.py`
5. Changed the old Hex command with a new menu based one.
6. The clear command can now take an optional member arg for check and its now menu based with confirmation.
7. The problem with the cuddle command has been fixed.
8. Husk now answers to users questions thanks to `celverbot api`
9. Move all command has become better in terms of error raising and member fetching.
10. added an SDK file for easier usage.
11. added functionality to the about command.

# Notes:
`bot is not using the stable version of d.py 1.6.0! its using a unstable clone from the master branch, its not recommended to use the master branch.`



# WillDo:
some of the things that are going to be added up to the version 3.0
1. guild based `Logger` for the bot with advanced logging dynamics
2. Full menu based bot configuration using the early version of `discord.Buttons`
3. refactoring the helpcommand to be more `Cat` based.
4. adding an `RolePlay` type game like `Mafia`.
5. [This is ridiculous] but using `discord.Buttons` to control a `gameboy emulator` via twitch
and many more..
so stay tuned :)

~David Marshal


ChangeLog.txt
Genral change : Added slash commands in cogs/slashcommands.py
Slash commands can currently only be used in guilds : 555067760703569960,801555851188895774

New SlashCommands are :
/hentai optional[hidden?]
/check user:onion[discord.User, discord.Member]
/moveall
/muteall
/unmuteall
/clear amount optional[user]
/gif category
--------------

1. changed the check command so now it wont show guild.default_role as one of the members roles

2. changed the unmuteall command so now it ignores the ctx.author

3. added getgifs func to the sdk cause im lazy :/

4. added slashguilds parameter to BotConfig() and bot.json

5. minor changes to some commands.

6. Took out the tenor api token from the bots gifs.py cog XD


main.py
from discord.ext import commands, tasks
from itertools import cycle
import discord.utils, discord, os
from sdk import husk_sdk
import async_cleverbot as ac
from discord_slash import SlashCommand

config = husk_sdk.BotConfig("bot.json")
config.Load()
intents = discord.Intents().all()
prefix = '!!'
status = cycle([f'{prefix}help', f'Version 2.0 Is Up', f'{prefix}Update to see it!',
                'Im UPDATED, Yaaaay!', 'slash commands are here!!'])
bot = commands.Bot(command_prefix=prefix, intents=intents, case_insensitive=True,
                   help_command=husk_sdk.HelpCommand())
slash = SlashCommand(bot, sync_commands=True, sync_on_cog_reload=True)
moods = {
    "joy": ac.Emotion.joy,
    "anger": ac.Emotion.anger,
    "angry": ac.Emotion.angry,
    "fear": ac.Emotion.fear,
    "sad": ac.Emotion.sad,
    "sadness": ac.Emotion.sadness,
    "happy": ac.Emotion.happy,
    "neutral": ac.Emotion.neutral,
    "normal": ac.Emotion.normal,
    "scared": ac.Emotion.scared
}
mood = moods["normal"]


@bot.event
async def on_ready():
    status_update.start()
    print('The bot has logged in as {0.user}\n-------------------↴'.format(bot))


# EVENT ZONE ----------↴
@bot.event
async def on_message(message):
    if message.content.lower().startswith("husk") and len(message.content[0:4]) > 3:
        await husk_sdk.call_CeleverBot(message, bot, mood)
    await bot.process_commands(message)


#######################
# main error handler #
#####################

@bot.event
async def on_command_error(message: discord.message, error):
    error = getattr(error, 'original', error)
    if isinstance(error, commands.CommandNotFound):
        await message.send(f"**ERROR** | `🚫` `Command Not Found, Use {prefix}help to see the bots commands`",
                           delete_after=3)
        await message.message.delete()
    else:
        errorlist = [commands.MissingPermissions,
                     commands.MissingRequiredArgument,
                     commands.UserNotFound,
                     commands.CommandInvokeError,
                     TimeoutError,
                     husk_sdk.NotInVoice,
                     husk_sdk.OverTheLimit,
                     commands.NotOwner]
        for er in errorlist:
            if isinstance(error, er):
                await message.send(f"**ERROR** | `🚫` `{error}`", delete_after=3)
                await message.message.delete()
                return
    print("ERROR:" + str(error))


@bot.event
async def on_raw_reaction_add(payload: discord.RawReactionActionEvent):
    role = payload.member.guild.get_role(807001981989814302)
    await payload.member.add_roles(role) if payload.message_id == 823641191567327235 else "x"


@bot.event
async def on_raw_reaction_remove(payload: discord.RawReactionActionEvent):
    guild: discord.Guild = bot.get_guild(payload.guild_id)
    member: discord.Member = guild.get_member(payload.user_id)
    role: discord.Role = guild.get_role(807001981989814302)
    await member.remove_roles(role) if payload.message_id == 823641191567327235 else "x"


# COMMAND ZONE ----------↴
@commands.is_owner()
@bot.command(hidden=True)
async def set_mood(ctx, inpmood: str):
    global mood
    try:
        mood = moods[inpmood]
        await ctx.send(f"**Done!, mood set to `{inpmood}`**", delete_after=5)
    except:
        await ctx.send("**Wrong Mood Given**", delete_after=5)

# LOOP ZONE ----------↴
@tasks.loop(minutes=2)
async def status_update():
    await bot.change_presence(activity=discord.Activity(type=discord.ActivityType.listening, name=status.__next__(),
                                                        url='https://discord.gg/Pr6qtxSUve'))


# Etc. ZONE ----------↴
for file in os.listdir('./cogs'):
    if file.endswith('.py'):
        bot.load_extension(f'cogs.{file[:-3]}')

bot.run(config.token)


cogs/gifs.py
import discord, json, requests, random
from discord.ext import commands
from sdk import husk_sdk
config = husk_sdk.BotConfig("./bot.json")
config.Load()
def user_check(user: discord.Member):
    if user is None:
        return "The air"
    if user is not None:
        return user.mention


class GifCommands(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
    #cat - dog - kiss - spank - lick - smile - greet - sleep - slap - cuddle
    @commands.command()
    async def cat(self, message):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('cat', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty, description=discord.Embed.Empty,
                                  color=discord.colour.Color.blue())
            embed.set_author(name="HUSK CAT GIFs",
                             icon_url='https://cdn6.aptoide.com/imgs/e/8/7/e87cd92ea75d17a681b1eef6b2b83670_icon.png')
            embed.set_image(url=gif['url'])
            embed.set_footer(text=f'size, about : {(gif["size"] / 1000).__round__()}KB')
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def dog(self, message):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('Dog', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty, description=discord.Embed.Empty,
                                  color=discord.colour.Color.blue())
            embed.set_author(name="HUSK DOG GIFs",
                             icon_url='https://cdn6.aptoide.com/imgs/e/8/7/e87cd92ea75d17a681b1eef6b2b83670_icon.png')
            embed.set_image(url=gif['url'])
            embed.set_footer(text=f'size, about : {(gif["size"] / 1000).__round__()}KB')
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def kiss(self, message, user :discord.Member=None):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('anime_kiss', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty, description=f"{message.author.name} **kissed** {user_check(user)}",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def slap(self, message, user: discord.Member = None):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('anime slap', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=f"{message.author.name} **slapped** {user_check(user)}",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def sleep(self, message):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('anime_kiss', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=f"{message.author.name} is going to sleep!",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def spank(self, message, user: discord.Member = None):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('anime spank', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=f"{message.author.name} **Spanked** {user_check(user)}",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def greet(self, message, user: discord.Member = None):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('aime greeting', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=f"{message.author.name} **greets** {user_check(user)}",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def smile(self, message, user: discord.Member = None):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('anime smile', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=f"{message.author.name} **smiled at** {user_check(user)}",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def lick(self, message, user: discord.Member = None):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('anime lick', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=f"{message.author.name} **licked** {user_check(user)}",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @commands.command()
    async def cuddle(self, message, user: discord.Member = None):
        first_msg = await message.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % ('anime cuddle', apikey, lmt))
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=f"{message.author.name} **Cuddled** {user_check(user)}",
                                  color=0x0000)
            embed.set_image(url=gif['url'])
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)
def setup(bot):
    bot.add_cog(GifCommands(bot))

cogs/personal.py
import discord
from discord.ext import commands
import requests

class PersonalCommands(commands.Cog):
        def __init__(self, bot):
            self.bot: commands.Bot = bot

        @commands.command(hidden=True)
        async def chaeyoung(self, message):

            if message.author.id == 467172810540843035:
                r = requests.api.get(f'https://meme-api.herokuapp.com/gimme/chaeyoung/3').json()
                z = []
                for x in r['memes']:
                    z.append(f'{x["url"]}')
                embed = discord.Embed(title=discord.Embed.Empty, description="-")
                embed.set_image(url=z[0])
                embed.set_thumbnail(url=z[1])
                await message.send(embed=embed)
            else:
                raise commands.MissingPermissions(["Chaeyoung's husband"])

def setup(bot):
    bot.add_cog(PersonalCommands(bot))
    print('personal.py loaded')

cogs/slashcommands.py
import discord, json, requests, random, webcolors
from discord.ext import commands
from discord_slash.utils.manage_commands import create_option, create_choice
from discord_slash import cog_ext, SlashContext
from sdk import husk_sdk

config = husk_sdk.BotConfig("./bot.json")
config.Load()

guild_ids = config.slashguilds


class SlashCommands(commands.Cog):
    def __init__(self, bot):
        self.bot: commands.Bot = bot

    @cog_ext.cog_slash(name="gif", description="gives you a gif from the options!", options=[
        create_option(name="category", description="Your predefined gif options", option_type=3, required=True,
                      choices=[create_choice(value=x, name=x) for x in husk_sdk.getgifs()])],
                       guild_ids=guild_ids)
    async def s_gif(self, ctx: SlashContext, category: str):
        first_msg = await ctx.send('**Retriving data**')
        apikey = config.gif_apikey
        lmt = 50
        try:
            r = requests.get("https://api.tenor.com/v1/search?q=%s&key=%s&limit=%s" % (category, apikey, lmt))
        except requests.exceptions.ConnectionError:
            return await ctx.send(f'**ERROR** | `🛑` `Something wet wrong, try again.`',
                                  delete_after=5)
        if r.status_code == 200:
            trending_gifs = json.loads(r.content)
            data = trending_gifs['results'][random.randint(0, trending_gifs['results'].__len__())]
            gif = None
            for nm in data['media']:
                if nm['gif']:
                    gif = nm['gif']
            embed = discord.Embed(title=discord.Embed.Empty, description=discord.Embed.Empty,
                                  color=discord.colour.Color.blue())
            embed.set_author(name="HUSK CAT GIFs",
                             icon_url='https://cdn6.aptoide.com/imgs/e/8/7/e87cd92ea75d17a681b1eef6b2b83670_icon.png')
            embed.set_image(url=gif['url'])
            embed.set_footer(text=f'size, about : {(gif["size"] / 1000).__round__()}KB')
            await first_msg.edit(content=None, embed=embed)
        else:
            await first_msg.edit(
                content=f'**ERROR** | `🛑` `Something wet wrong, try again. Respond-Code :{r.status_code}`',
                delete_after=5)

    @cog_ext.cog_slash(guild_ids=guild_ids, name="clear",
                       description="Deletes the amount of messages givin to it based on the user, default is global purge!",
                       options=[create_option(name="amount", description="Amount of messages to delete!", option_type=4,
                                              required=True),
                                create_option(name="user", description="specific user to check", option_type=6,
                                              required=False)])
    async def s_clear(self, ctx: SlashContext, amount: int, user: discord.Member = None):
        if not amount <= 100 and amount >= 1:
            raise husk_sdk.OverTheLimit("1 -> 100", amount)
        else:
            check = lambda message: (message.author == user) if user is not None else True
            await ctx.channel.purge(limit=amount, check=check)
            embed = discord.Embed(title=f"**DONE :-)**",
                                  description=f"```dif\nThe Amount of {amount} messages have been purged from" \
                                              f" {ctx.channel.name}```" \
                                      if user is None else f"```dif\nThe Amount of {amount} messages from {user.display_name}" \
                                                           f" have been purged from {ctx.channel.name}```")
            embed.set_footer(text=f"Used by 🔹{ctx.author.display_name}🔹"
                             , icon_url=ctx.author.avatar.url)
            await ctx.send(embed=embed)

    @cog_ext.cog_slash(name="moveall", description="Moves all the members in a server, to authors VoiceChannel!",
                       guild_ids=guild_ids)
    async def s_moveall(self, ctx: SlashContext):
        toMove = [mem for voice in [voice.members for voice in ctx.guild.voice_channels
                                    if voice.members is not None] for mem in
                  voice]  ##gets a list of every member in a VC
        moved = []
        for member in toMove:
            if member != ctx.author and member not in ctx.author.voice.channel.members:
                await member.move_to(ctx.author.voice.channel)
                moved.append(member)
        embed = discord.Embed(title=f"Moved `{len(moved)} Members` to `{ctx.author.voice.channel.name}`",
                              description="**__Moved Members__**:\n" +
                                          "\n".join([member.mention for member in moved]), color=discord.Color.random()) \
            .set_footer(text=f"Used by 🔹{ctx.author.display_name}🔹",
                        icon_url=ctx.author.avatar.url) if moved != [] else \
            discord.Embed(title=f"**Moveall command**", description="**🚫 No one is in other VoiceChannels!**",
                          color=discord.Color.red())
        await ctx.send(embed=embed, delete_after=None if moved != [] else 10)

    @cog_ext.cog_slash(name="muteall", description="Mutes all the members in authors VC!",
                       guild_ids=guild_ids)
    async def s_muteall(self, ctx: commands.Context):
        muted = []
        for member in ctx.author.voice.channel.members:
            if member != ctx.author:
                await member.edit(mute=True)
                muted.append(member)
        embed = discord.Embed(title=f"Muted `{len(muted)} Members` in `{ctx.author.voice.channel.name}`",
                              description="**__Muted Members__**:\n" +
                                          "\n".join([member.mention for member in muted]), color=discord.Color.random()) \
            .set_footer(text=f"Used by 🔹{ctx.author.display_name}🔹",
                        icon_url=ctx.author.avatar.url) if muted != [] else \
            discord.Embed(title=f"**Muteall command**", description="**🚫 No one is in this VoiceChannel except you!**",
                          color=discord.Color.red())
        await ctx.send(embed=embed, delete_after=None if muted != [] else 10)

    @cog_ext.cog_slash(name="unmuteall", description="Unmutes all the members in authors VC!", guild_ids=guild_ids)
    async def s_unmuteall(self, ctx: commands.Context):
        unmuted = []
        for member in ctx.author.voice.channel.members:
            if member.voice.mute:
                if member != ctx.author:
                    await member.edit(mute=False)
                    unmuted.append(member)
        embed = discord.Embed(title=f"UnMuted `{len(unmuted)} Members` in `{ctx.author.voice.channel.name}`",
                              description="**__UnMuted Members__**:\n" +
                                          "\n".join([member.mention for member in unmuted]),
                              color=discord.Color.random()) \
            .set_footer(text=f"Used by 🔹{ctx.author.display_name}🔹",
                        icon_url=ctx.author.avatar.url) if unmuted != [] else \
            discord.Embed(title=f"**UnMuteall command**",
                          description="**🚫 No one is in this VoiceChannel except you!**",
                          color=discord.Color.red())
        await ctx.send(embed=embed, delete_after=None if unmuted != [] else 10)

    @cog_ext.cog_slash(name="check", description="Gives you info about an specific member anonymously",
                        options=[create_option(name="user", description="specific user to check", option_type=6,
                                              required=True)], guild_ids=guild_ids)
    async def s_check(self, ctx: SlashContext, user: discord.User):
        member: discord.Member = discord.utils.get(ctx.guild.members, id=user.id)
        try:
            color_web = webcolors.name_to_rgb(f'{user.default_avatar}')
            color = discord.Colour.from_rgb(color_web.red, color_web.green, color_web.blue)
        except ValueError:
            color = discord.Color.random()
        embed = discord.Embed(title=f"User's DisplayName: `{user.display_name}`",
                              description=f"**User's id** : `{user.id}`\n"
                                          f"**User's Discriminator** : `#{user.discriminator}`",
                              colour=color)

        embed.set_thumbnail(url=user.avatar.url)
        embed.set_footer(text=f"Used by 🔹{ctx.author.display_name}🔹",
                         icon_url=ctx.author.avatar.url)
        embed.add_field(name="`Account Specific Detail`**:**",
                        value=f"""```Account Created in : {user.created_at.strftime('%Y-%m-%d')}
Animated Profile Avatar? : {'Yes' if user.avatar.is_animated() else 'No'}
is a Bot? : {'Yes' if user.bot else 'No'}
Default Avatar Color : {user.colour}
Current Activity : {member.activity.name if member.activity is not None else 'Nothing'}
is on mobile? : {'Yes' if member.is_on_mobile() else 'No'}
Status : {'Do not Disturb' if member.status == 'dnd' else member.status}```""")
        embed.add_field(name=f"`Server Specific Detail`**:**",
                        value=f"""```current Server : {ctx.guild.name}
NickName : {'None' if member.nick is None else member.nick}
joined the server at : {member.joined_at.strftime("%Y-%m-%d")}
Top Role : {member.top_role.name}
Is Boosting the server? : {'Yes' if member.premium_since is not None else 'No'}
Server Roles : {"-".join([role.name for role in member.roles if role.name != "@everyone"])}
is Server Owner : {'Yes' if member == ctx.guild.owner else 'No'}
                                ```""",
                        inline=False)
        await ctx.send(embed=embed, hidden=True)
    @cog_ext.cog_slash(name="hentai", description="Gives you some good stuff from r/hentai", options=[
        create_option(name="hidden", description="Only you can see the result if `YES`", required=False, option_type=4, choices=[
            create_choice(name="YES, just for me", value=1), create_choice(name="NO, lets enjoy it together", value=0)])], guild_ids=guild_ids)
    async def s_hentai(self, ctx: SlashContext, hidden: int = 0):
        hidden = bool(hidden)
        def embed_msg(data):
            embed = discord.Embed(title=discord.Embed.Empty,
                                  description=discord.Embed.Empty,
                                  color=discord.colour.Color.dark_orange())
            embed.set_author(name=f"{data['title']}", url=data['postLink'],
                             icon_url='https://2.bp.blogspot.com/-r3brlD_9eHg/XDz5bERnBMI/AAAAAAAAG2Y/XfivK0eVkiQej2t-xfmlNL6MlSQZkvcEACK4BGAYYCw/s1600/logo%2Breddit.png')
            embed.set_image(url=data['url'])
            embed.set_footer(
                text=f'⬆ UpVotes : {data["ups"]} , subreddit : {data["subreddit"]},\n 🔞 NSFW : {data["nsfw"]}, ✍author : {data["author"]}')
            return embed

        r = requests.api.get(f'https://meme-api.herokuapp.com/gimme/hentai')
        if r.status_code == 200:
            data = r.json()
            if data['nsfw']:
                if ctx.channel.is_nsfw():
                    await ctx.send(embed=embed_msg(data), hidden=hidden)
                else:
                    await ctx.send(f'**NSFW** | `🔞` `{ctx.channel.name} is not an NSFW channel!`', hidden=hidden)
            else:
                await ctx.send(embed=embed_msg(data), hidden=hidden)
        else:
            await ctx.send(f'**REDDIT ERROR** | `🚫` `{r.json()}`')
def setup(bot):
    bot.add_cog(SlashCommands(bot))
    print('slashcommands.py loaded')
